### PR TITLE
suggest `crate::...` for "local" paths in 2018

### LINKED ISSUE #54230: suggest `crate::...` for "local" paths in 2018
The following code, [when compiled in Rust 2018](https://play.rust-lang.org/?gist=100abed41bd0a5cd89ecc7ee7d3a8e82&version=nightly&mode=debug&edition=2018), errors out (as expected). However, the error could be more helpful.

```rust
mod foo {
    type Bar = u32;
}

use foo::Bar;

fn main() {
    let x: Bar = 22;
}
```

Error:

```
error[E0463]: can't find crate for `foo`
 --> src/main.rs:5:5
  |
5 | use foo::Bar;
  |     ^^^ can't find crate

error: aborting due to previous error
```

but I think we should suggest something like

```
did you mean `crate::foo::Bar`?
```

### PR DESCRIPTION
**Pull request by davidtwco**
Fixes #54230.

This commit adds suggestions for unresolved imports in the cases where
there could be a missing `crate::`, `super::`, `self::` or a missing
external crate name before an import.

r? @nikomatsakis

### CODE REVIEW

**Review by nikomatsakis — COMMENTED** [review]
Hmm, nice!

**Comment by nikomatsakis** [comment]
The code looks pretty good to me. I think there are some policy questions, e.g. the way that it prefers `self` or `super` to `crate`, that are not entirely clear to me. But probably not a big deal in practice.

**Comment by nikomatsakis** [comment]
Or, I guess that the current heuristic is:

- `self::foo`
- `crate::foo`
- `super::bar`
- `extern_crate::bar`

It feels like it might be reasonable to show more than one, too, though it'd be lame to do so if they are the same thing underneath.

**Comment by davidtwco** [comment]
> 
> 
> The code looks pretty good to me. I think there are some policy questions, e.g. the way that it prefers `self` or `super` to `crate`, that are not entirely clear to me. But probably not a big deal in practice.

Yeah, I've ordered these entirely arbitrarily - we can adjust that. I also experimented with having it attempt to correct misspellings but I couldn't quite tweak it so it wasn't over- or under-correcting. I would get it so if I had `tsd::path` it would suggest `std::path` but it would also suggest that for things that were completely wrong; and on the other end of the scale, if I had a long path that had one or two characters swapped, it wouldn't suggest anything.

**Comment by nikomatsakis** [comment]
I'm feeling a bit confused now because I know there is also this  `lookup_import_candidates_from_module` function, and I'm trying to figure out how/if these ought to relate to one another. =)

**Comment by davidtwco** [comment]
@nikomatsakis If it is at all helpful, these are the rough notes I threw together when initially working out what was happening and where these errors were coming from:

<details>
Searching for the error message, we find it in `resolve_path_with_parent_scope`:

https://github.com/rust-lang/rust/blob/186fe091434b4c20c160b8098a56bb6a841bf6b1/src/librustc_resolve/lib.rs#L3673-L3689

This returns a `PathResult::Failed(..)` with a span, message and boolean indicating if this is the last error. `resolve_path_with_parent_scope` is called by `resolve_path`.

`resolved_path` is called by `resolve_import` where the `PathResult` is matched, returning `true`:

https://github.com/rust-lang/rust/blob/186fe091434b4c20c160b8098a56bb6a841bf6b1/src/librustc_resolve/resolve_imports.rs#L876-L880

`resolve_import` is called by `resolved_imports` where the `true` is matched against to add the import that failed to the `determined_imports` list:

https://github.com/rust-lang/rust/blob/186fe091434b4c20c160b8098a56bb6a841bf6b1/src/librustc_resolve/resolve_imports.rs#L628-L631

Later, `finalize_imports` is called, this iterates over the `determined_imports` which calls `finalize_import`:

https://github.com/rust-lang/rust/blob/186fe091434b4c20c160b8098a56bb6a841bf6b1/src/librustc_resolve/resolve_imports.rs#L657-L659

`finalize_import` eventually returns a span and a message, this is where we can see the original `self::` suggestion being made:

https://github.com/rust-lang/rust/blob/186fe091434b4c20c160b8098a56bb6a841bf6b1/src/librustc_resolve/resolve_imports.rs#L960-L973

This then goes back to `finalize_imports` which calls adds the span and message to a vector:

https://github.com/rust-lang/rust/blob/186fe091434b4c20c160b8098a56bb6a841bf6b1/src/librustc_resolve/resolve_imports.rs#L728

That vector then is passed to `throw_unresolved_import_error`:

https://github.com/rust-lang/rust/blob/186fe091434b4c20c160b8098a56bb6a841bf6b1/src/librustc_resolve/resolve_imports.rs#L805

`throw_unresolved_import_error` throws the actual error that we see:

https://github.com/rust-lang/rust/blob/186fe091434b4c20c160b8098a56bb6a841bf6b1/src/librustc_resolve/resolve_imports.rs#L821
</details>

File: src/librustc_resolve/error_reporting.rs [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        // Need to clone else we can't call `resolve_path` without a borrow error.
+        let external_crate_names = self.extern_prelude.clone();
+
+        // Insert a new path segment that we can replace.
+        let new_path_segment = path[0].clone();
+        path.insert(1, new_path_segment);
+
+        for name in &external_crate_names {
+            // Don't suggest meta as it will error in `resolve_path`.
+            if name.as_str() == "meta" {
```
Comment (eddyb) [inline src/librustc_resolve/error_reporting.rs]: Hardcoding this is wrong - you want to use something like this instead: https://github.com/rust-lang/rust/blob/c222479c6feb036e5e5198e068aa9824d855ffa4/src/librustc_resolve/lib.rs#L4403-L4407

**Comment by eddyb** [comment]
I think this PR might need to check the edition? Or somehow involve the compat lints.
That said,
r? @petrochenkov

**Comment by davidtwco** [comment]
@eddyb I've added a test demonstrating the 2015 behaviour for the same test case.

**Pull request merged**